Unable to build executable using embedded zip Python

## Description of the issue
The official Python embeddable zip file distributions for Windows from Python.org contain a zip file (pythonXX.zip) that contains the pre-compiled and optimized .pyc files to run Python isolated on a machine. I've installed the embedded Python on a local agent that we use in our azure devops pipeline, following the [Microsoft docs](https://docs.microsoft.com/en-us/azure/devops/pipelines/tasks/tool/use-python-version?view=azure-devops#how-can-i-configure-a-self-hosted-agent-to-use-this-task).

However, building on this agent using PyInstaller fails, see Stacktrace below. This failure occurs because the requirements for `base_library.zip` are located inside a zip. Thus, PyInstaller tries to do an `os.stat()` on a path inside a zip folder, which fails.

## Context information (for bug reports)

* Output of `pyinstaller --version`: ```3.6```
* Version of Python: 3.7.7
* Platform: Windows 10
* Same failure occurs on newest development version of PyInstaller
* Followed all the instructions in the "If Things Go Wrong" Guide and checked that everything is packaged correctly. The error occurs regardless of PyInstaller settings or spec file content.

### A minimal example program which shows the error
- Download embedded Python from [Python.org](https://www.python.org/downloads/)
- Add pip using [get-pip](https://michlstechblog.info/blog/python-install-python-with-pip-on-windows-by-the-embeddable-zip-file/)
- Install PyInstaller
- Create `helloworld.py`
   ```
   print("Hello World!")
   ```
- Try to run PyInstaller on `helloworld.py` using any settings.

### Stacktrace / full error message
```
==============================================================================
Task         : Command line
Description  : Run a command line script using Bash on Linux and macOS and cmd.exe on Windows
Version      : 2.164.2
Author       : Microsoft Corporation
Help         : https://docs.microsoft.com/azure/devops/pipelines/tasks/utility/command-line
==============================================================================
Generating script.
Script contents:
python -m PyInstaller --distpath "%cd%\exe" SPECFILENAME.spec
========================== Starting Command Output ===========================
"C:\WINDOWS\system32\cmd.exe" /D /E:ON /V:OFF /S /C "CALL "agent2\_work\_temp\44e0b0b7-f2f9-4990-a3ee-31bbec9a0c70.cmd""
285 INFO: PyInstaller: 3.6
285 INFO: Python: 3.7.7
285 INFO: Platform: Windows-10-10.0.18362-SP0
291 INFO: UPX is not available.
303 INFO: checking Analysis
307 INFO: Building Analysis because Analysis-00.toc is non existent
307 INFO: Initializing module dependency graph...
313 INFO: Caching module graph hooks...
326 INFO: Analyzing base_library.zip ...
1791 INFO: Processing pre-find module path hook   distutils
1804 INFO: distutils: retargeting to non-venv dir 'agent2\\_work\\_tool\\Python\\3.7.7\\x64\\python37.zip'
2579 INFO: Caching module dependency graph...
2629 ERROR: base_library.zip could not be created!
Traceback (most recent call last):
  File "D:\obj\windows-release\37amd64_Release\msi_python\zip_amd64\runpy.py", line 193, in _run_module_as_main
  File "D:\obj\windows-release\37amd64_Release\msi_python\zip_amd64\runpy.py", line 85, in _run_code
  File "agent2\_work\_tool\Python\3.7.7\x64\lib\site-packages\PyInstaller\__main__.py", line 121, in <module>
    run()
  File "agent2\_work\_tool\Python\3.7.7\x64\lib\site-packages\PyInstaller\__main__.py", line 114, in run
    run_build(pyi_config, spec_file, **vars(args))
  File "agent2\_work\_tool\Python\3.7.7\x64\lib\site-packages\PyInstaller\__main__.py", line 65, in run_build
    PyInstaller.building.build_main.main(pyi_config, spec_file, **kwargs)
  File "agent2\_work\_tool\Python\3.7.7\x64\lib\site-packages\PyInstaller\building\build_main.py", line 734, in main
    build(specfile, kw.get('distpath'), kw.get('workpath'), kw.get('clean_build'))
  File "agent2\_work\_tool\Python\3.7.7\x64\lib\site-packages\PyInstaller\building\build_main.py", line 681, in build
    exec(code, spec_namespace)
  File "SPECFILENAME.spec", line 24, in <module>
    noarchive=False)
  File "agent2\_work\_tool\Python\3.7.7\x64\lib\site-packages\PyInstaller\building\build_main.py", line 244, in __init__
    self.__postinit__()
  File "agent2\_work\_tool\Python\3.7.7\x64\lib\site-packages\PyInstaller\building\datastruct.py", line 160, in __postinit__
    self.assemble()
  File "agent2\_work\_tool\Python\3.7.7\x64\lib\site-packages\PyInstaller\building\build_main.py", line 352, in assemble
    create_py3_base_library(libzip_filename, graph=self.graph)
  File "agent2\_work\_tool\Python\3.7.7\x64\lib\site-packages\PyInstaller\depend\utils.py", line 74, in create_py3_base_library
    st = os.stat(mod.filename)
FileNotFoundError: [WinError 3] The system cannot find the path specified: 'agent2\\_work\\_tool\\Python\\3.7.7\\x64\\python37.zip\\encodings\\__init__.pyc'
##[error]Cmd.exe exited with code '1'.
```

--- a/PyInstaller/building/utils.py
+++ b/PyInstaller/building/utils.py
@@ -24,6 +24,7 @@ import shutil
 import struct
 import subprocess
 import sys
+import zipfile
 
 from PyInstaller import compat
 from PyInstaller import log as logging
@@ -593,22 +594,38 @@ def get_code_object(modname, filename):
     This is a simplifed non-performant version which circumvents __pycache__.
     """
 
-    try:
-        if filename in ('-', None):
-            # This is a NamespacePackage, modulegraph marks them by using the filename '-'. (But wants to use None, so
-            # check for None, too, to be forward-compatible.)
-            logger.debug('Compiling namespace package %s', modname)
-            txt = '#\n'
-            return compile(txt, filename, 'exec')
+    if filename in ('-', None):
+        # This is a NamespacePackage, modulegraph marks them by using the filename '-'. (But wants to use None, so
+        # check for None, too, to be forward-compatible.)
+        logger.debug('Compiling namespace package %s', modname)
+        txt = '#\n'
+        code_object = compile(txt, filename, 'exec')
+    else:
+        _, ext = os.path.splitext(filename)
+        ext = ext.lower()
+
+        if ext == '.pyc':
+            # The module is available in binary-only form. Read the contents of .pyc file using helper function, which
+            # supports reading from either stand-alone or archive-embedded .pyc files.
+            logger.debug('Reading code object from .pyc file %s', filename)
+            pyc_data = _read_pyc_data(filename)
+            code_object = marshal.loads(pyc_data[16:])
         else:
-            logger.debug('Compiling %s', filename)
+            # Assume this is a source .py file, but allow an arbitrary extension (other than .pyc, which is taken in
+            # the above branch). This allows entry-point scripts to have an arbitrary (or no) extension, as tested by
+            # the `test_arbitrary_ext` in `test_basic.py`.
+            logger.debug('Compiling python script/module file %s', filename)
+
             with open(filename, 'rb') as f:
                 source = f.read()
-            return compile(source, filename, 'exec')
-    except SyntaxError as e:
-        print("Syntax error in ", filename)
-        print(e.args)
-        raise
+
+            try:
+                code_object = compile(source, filename, 'exec')
+            except SyntaxError:
+                logger.warning("Sytnax error while compiling %s", filename)
+                raise
+
+    return code_object
 
 
 def strip_paths_in_code(co, new_filename=None):
@@ -722,12 +739,9 @@ def compile_pymodule(name, src_path, workpath, code_cache=None):
             with open(pyc_path, 'rb') as fp:
                 pyc_data = fp.read()
         elif ext == '.pyc':
-            # The module is available in binary-only form. Read it...
-            with open(src_path, 'rb') as fp:
-                pyc_data = fp.read()
-            # ... verify the python version...
-            if pyc_data[:4] != compat.BYTECODE_MAGIC:
-                raise ValueError(f"The .pyc module {src_path} was compiled for incompatible version of python!")
+            # The module is available in binary-only form. Read the contents of .pyc file using helper function, which
+            # supports reading from either stand-alone or archive-embedded .pyc files.
+            pyc_data = _read_pyc_data(src_path)
         else:
             raise ValueError(f"Invalid python module file {src_path}; unhandled extension {ext}!")
 
@@ -746,6 +760,35 @@ def compile_pymodule(name, src_path, workpath, code_cache=None):
 
     # Return output path
     return pyc_path
+
+
+def _read_pyc_data(filename):
+    """
+    Helper for reading data from .pyc files. Supports both stand-alone and archive-embedded .pyc files. Used by
+    `compile_pymodule` and `get_code_object` helper functions.
+    """
+    src_file = pathlib.Path(filename)
+
+    if src_file.is_file():
+        # Stand-alone .pyc file.
+        pyc_data = src_file.read_bytes()
+    else:
+        # Check if .pyc file is stored in a .zip archive, as is the case for stdlib modules in embeddable
+        # python on Windows.
+        parent_zip_file = misc.path_to_parent_archive(src_file)
+        if parent_zip_file is not None and zipfile.is_zipfile(parent_zip_file):
+            with zipfile.ZipFile(parent_zip_file, 'r') as zip_archive:
+                # NOTE: zip entry names must be in POSIX format, even on Windows!
+                zip_entry_name = str(src_file.relative_to(parent_zip_file).as_posix())
+                pyc_data = zip_archive.read(zip_entry_name)
+        else:
+            raise FileNotFoundError(f"Cannot find .pyc file {filename!r}!")
+
+        # Verify the python version
+        if pyc_data[:4] != compat.BYTECODE_MAGIC:
+            raise ValueError(f"The .pyc module {filename} was compiled for incompatible version of python!")
+
+    return pyc_data
 
 
 def postprocess_binaries_toc_pywin32(binaries):

--- a/PyInstaller/utils/misc.py
+++ b/PyInstaller/utils/misc.py
@@ -19,6 +19,7 @@ import codecs
 import re
 import tokenize
 import io
+import pathlib
 
 from PyInstaller import log as logging
 from PyInstaller.compat import is_win
@@ -232,3 +233,15 @@ def is_iterable(arg):
     except TypeError:
         return False
     return True
+
+
+def path_to_parent_archive(filename):
+    """
+    Check if the given file path points to a file inside an existing archive file. Returns first path from the set of
+    parent paths that points to an existing file, or `None` if no such path exists (i.e., file is an actual stand-alone
+    file).
+    """
+    for parent in pathlib.Path(filename).parents:
+        if parent.is_file():
+            return parent
+    return None

--- a/news/4989.feature.rst
+++ b/news/4989.feature.rst
@@ -1,0 +1,2 @@
+(Windows) Add support for collecting .pyc files from the ``python3X.zip`` archive
+where ``Windows embeddable package`` Python stores its stdlib modules.

